contributors/design-proposals/vertical-pod-autoscaler.md

```diff
@@ -0,0 +1,486 @@
+1. Downtime of heavy-weight components (database/recommender) must not block
+   recreating existing Pods. Components on critical path for Pod creation
+   (admission controller) are designed to be highly available.
+
+### Extensibility ###
+1. VPA is capable of performing in-place updates once they become available.
+
+------
+Design
+------
+
+### Overview ###
+(see further sections for details and justification)
+
+1. We introduce a new type of **API resource**:
+   `VerticalPodAutoscaler`. It consists of a **label selector** to match Pods,
+   the **resources policy** (controls how VPA computes the resources), the
+   **update policy** (controls how changes are applied to Pods) and the
+   recommended Pod resources (an output field).
+   
+2. **VPA Recommender** is a new component which **consumes utilization signals
+   and OOM events** for all Pods in the cluster from Heapster (or another
+   monitoring solution).
+   
+3. VPA Recommender **watches all Pods**, keeps calculating fresh recommended
+   resources for them and **stores the recommendations in the VPA objects**.
+   
+4. Additionally the Recommender **exposes a synchronous API** that takes a Pod
+   description and returns recommended resources.
+   
+5. All Pod creation requests go through the VPA **Admission Controller**.
+   If the Pod is matched by any VerticalPodAutoscaler object, the admission
+   controller **overrides resources** of containers in the Pod with the
+   recommendation provided by the VPA Recommender. If the Recommender is not
+   available, it falls back to the recommendation cached in the VPA object.
+   
+6. Pods can use the [**Initializer**](https://github.com/kubernetes/kubernetes/issues/3585)
+   mechanism to ensure they are not scheduled until a matching VPA object is
+   present. This is a recommended way of using VPA.
+   
+7. **VPA Updater** is a component responsible for **real-time updates** of Pods.
+   If a Pod uses VPA in `"auto"` mode, the Updater can decide to update it with
+   recommender resources.
+   In MVP this is realized by just evicting the Pod in order to have it
+   recreated with new resources. This approach requires the Pod to belong to a
+   Replica Set (or some other owner capable of recreating it).
+   In future the Updater will take advantage of in-place updates, which would
+   most likely lift this constraint.
+   Because restarting/rescheduling Pods is disruptive to the service, it must be
+   rare.
+   
+8. VPA only controls the resource **request** of containers. It sets the limit
+   to infinity. The request is calculated based on analysis of the current and
+   previous runs (see [Recommendation model](#recommendation-model) below).
+   
+9. **History Storage** is a component that consumes utilization signals and OOMs
+   (same data as the Recommender) from Heapster and stores it persistently.
+   It is used by the Recommender to **initialize its state on startup**.
+   First implementation will be InFluxDB.
```

Comment:
AFAIR Influxdb is currently installed by default on OSS GCE cluster which is not true for Prometheus that requires a bunch of extra steps to be enabled in the first place. So for the initial alpha version we will focus on influxdb. 
In the longer run we want to define an API and allow multiple time series databases, including Prometheus, to work with VPA.